### PR TITLE
Extend intelmq_psql_initdb

### PR DESCRIPTION
**Pull request by kamil-certat**
Added creating view for separating raws, optional partitioning key, 
custom harmonization file path and `if not exists`/`or replace` clauses.

Those changes improves the capabilities of the database
schema creation. It's now easier to switch to the TimescaleDB
or use partitioning in PostgreSQL. Creating scripts can be
generated from the harmonisation, as previously just the most
basic version. If requested, creation commands get IF NOT EXISTS
or OR REPLACE directives to allow re-runing script.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ CHANGELOG
 ### Tests
 
 ### Tools
+ - `intelmq_psql_initdb` got support for providing custom harmonization file, generating view for storing `raw` fields separately, and adding `IF NOT EXISTS`/`OR REPLACE` clauses (PR by Kamil Mankowski).
```
Comment (sebix) [inline CHANGELOG.md]:
```suggestion
 - `intelmq_psql_initdb` got support for providing custom harmonization file, generating view for storing `raw` fields separately, and adding `IF NOT EXISTS`/`OR REPLACE` clauses ([PR#2404](https://github.com/certtools/intelmq/pull/2404) by Kamil Mankowski).
```

File: intelmq/bin/intelmq_psql_initdb.py [hunk 2]
```diff
@@ -97,6 +199,18 @@ def main():
...
                         default='/tmp/initdb.sql'
                         )
+    parser.add_argument("--no-events", action="store_true", default=False,
+                        help="Skip generating the events table schema")
+    parser.add_argument("--separate-raws", action="store_true", default=False,
+                        help="Generate v_events view to separate raw field from the rest of the data on insert")
+    parser.add_argument("--partition-key", default=None,
+                        help=("Add given field to the primary key of the events table. "
+                              "Useful when utilizing partitioning for TimescaleDB. "
+                              "If combined with --separate-raws, the v_events does not get foreign key"))
```
Comment (sebix) [inline intelmq/bin/intelmq_psql_initdb.py]: From a user perspective this information is unfortunately not enough to comprehend what this option is used for and how to use it best. Is `time.source` a good option, so also use https://intelmq.readthedocs.io/en/develop/user/eventdb.html#using-eventdb-with-timescale-db ?

**Review by sebix — CHANGES_REQUESTED** [review]
Great to see that my design is still useful :)

Is the file `contrib/eventdb/separate-raws-table.sql` obsoleted?

https://intelmq.readthedocs.io/en/develop/user/eventdb.html#separating-raw-values-in-postgresql-using-view-and-trigger needs an update for the new method.

**Comment by kamil-certat** [comment]
Thanks for catching lack of docs, let's see now

File: docs/user/eventdb.rst [hunk 3]
```diff
@@ -8,22 +8,42 @@ EventDB
 
 The EventDB is not a software itself.
 
-The EventDB is a database (usually `PostgreSQL <postgresql.org/>`_) that gets filled with with data from IntelMQ using the :ref:`intelmq.bots.outputs.sql.output` Output Bot.
+The EventDB is a database (usually `PostgreSQL <postgresql.org/>`_) that gets filled with with data
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
The EventDB is a database (usually `PostgreSQL <postgresql.org/>`_) that gets filled with data
```

File: docs/user/eventdb.rst [hunk 4]
```diff
@@ -8,22 +8,42 @@ EventDB
...
+
+In addition, the script supports some additional features supporting use cases described later in
+this document:
+
+- ``--partition-key`` - for generating schema aligned with :ref:`TimescaleDB <timescaledb>`
+  or partitioned tables,
+- ``--separate-raws`` - for generating views and triggers needed to :ref:`eventdb_raws_table`
+  (works also together with adjustments for partitioning).
+
+For full list of supported parameters, call the script help using ``-h`` parameter.
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
For a full list of supported parameters, call the script help using ``-h`` parameter.
```

File: docs/user/eventdb.rst [hunk 5]
```diff
@@ -8,22 +8,42 @@ EventDB
...
 
-Having an `events` table as outlined in the SQL file, IntelMQ's :ref:`intelmq.bots.outputs.sql.output` Output Bot can write all received events into this database table.
+All elements of generated SQL file can be adapted and extended before running the SQL file against
+a database, especially the indexes. Please review the generated script before applying.
 
-This events table is the core of the so-called EventDB and also required by all other sections of this document.
+Be aware that if you create tables using another DB user that used later by the output bot, you may
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
Be aware that if you create tables using another DB user that is used later by the output bot, you may
```

File: docs/user/eventdb.rst [hunk 6]
```diff
@@ -8,22 +8,42 @@ EventDB
...
+
+- ``--partition-key`` - for generating schema aligned with :ref:`TimescaleDB <timescaledb>`
+  or partitioned tables,
+- ``--separate-raws`` - for generating views and triggers needed to :ref:`eventdb_raws_table`
+  (works also together with adjustments for partitioning).
+
+For full list of supported parameters, call the script help using ``-h`` parameter.
 
-Having an `events` table as outlined in the SQL file, IntelMQ's :ref:`intelmq.bots.outputs.sql.output` Output Bot can write all received events into this database table.
+All elements of generated SQL file can be adapted and extended before running the SQL file against
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
All elements of the generated SQL file can be adapted and extended before running the SQL file against
```

File: docs/user/eventdb.rst [hunk 7]
```diff
@@ -91,10 +113,28 @@ Time-series has been invented as traditional database design like relational or
...
+------------------------------
+
+To utilize the time-series, you need to choose a column containing the right time. This is then
+used by you for manual queries and graphs, but also by the database itself for organizing the data.
+
+The :doc:`IntelMQ Data Format </dev/harmonization-fields>` has two fields that can be used for this:
+``time.source`` or ``time.observation``. Depending on your needs (tracking when event occurred or when
+was detected, if different) choose one of them.
+
+You can use :ref:`intelmq_psql_initdb` tool to generate SQL schema valid for TimescaleDB by passing
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
You can use the :ref:`intelmq_psql_initdb` tool to generate SQL schema valid for TimescaleDB by passing
```

File: docs/user/eventdb.rst [hunk 8]
```diff
@@ -91,10 +113,28 @@ Time-series has been invented as traditional database design like relational or
...
+the partitioning key:
+
+.. code-block:: bash
+
+   intelmq_psql_initdb --partition-key "time.source"
+
+How to setup?
+-------------
+
+Thanks to TimescaleDB it's very easy to setup.
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
Thanks to TimescaleDB, it's very easy to setup.
```

File: docs/user/eventdb.rst [hunk 9]
```diff
@@ -91,10 +113,28 @@ Time-series has been invented as traditional database design like relational or
...
-How to setup
-------------
+How to choose the time column?
+------------------------------
+
+To utilize the time-series, you need to choose a column containing the right time. This is then
+used by you for manual queries and graphs, but also by the database itself for organizing the data.
+
+The :doc:`IntelMQ Data Format </dev/harmonization-fields>` has two fields that can be used for this:
+``time.source`` or ``time.observation``. Depending on your needs (tracking when event occurred or when
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
``time.source`` or ``time.observation``. Depending on your needs (tracking when the event occurred or when
```

File: docs/user/eventdb.rst [hunk 10]
```diff
@@ -91,10 +113,28 @@ Time-series has been invented as traditional database design like relational or
...
+was detected, if different) choose one of them.
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
it was detected, if different), choose one of them.
```

File: docs/user/eventdb.rst [hunk 11]
```diff
@@ -91,10 +113,28 @@ Time-series has been invented as traditional database design like relational or
 A big benefit of time-series instead of other database designs over a time-based search pattern is the performance.
 As IntelMQ uses data based upon time, this design is awesome & will give you a performance boost.
 
-How to setup
-------------
+How to choose the time column?
+------------------------------
+
+To utilize the time-series, you need to choose a column containing the right time. This is then
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
To utilize the time-series, choose a column containing the right time. This is then
```

File: docs/user/eventdb.rst [hunk 12]
```diff
@@ -91,10 +113,28 @@ Time-series has been invented as traditional database design like relational or
...
+used by you for manual queries and graphs, but also by the database itself for organizing the data.
```
Comment (sebix) [inline docs/user/eventdb.rst]:
```suggestion
used by you for manual queries and graphs, and also by the database itself for organizing the data.
```

**Review by sebix — APPROVED** [review]
Only wording suggestions. LGTM

**Comment by kamil-certat** [comment]
@sebix Would you find time to merge this PR?

**Pull request merged**